File: src/gateway/connections.ts
/**
 * 连接管理域的 gateway 封装:命令名与入参形状对齐 Rust commands 层
 * (TECHNICAL_DESIGN §6.2)。全部经 invokeUnwrapped 解包,失败抛 GatewayError。
 */
import type {
  ConnectionDto,
  ConnectionInput,
  ConnectionNodeDto,
  DeleteGroupMode,
  GroupDto,
} from "@/types";
import { invokeUnwrapped } from "./tauri";

/** 列出连接树(分组在前、连接在后,同级已排序)。 */
export function listConnections(): Promise<ConnectionNodeDto[]> {
  return invokeUnwrapped("list_connections");
}

/** 新建连接。 */
export function createConnection(input: ConnectionInput): Promise<ConnectionDto> {
  return invokeUnwrapped("create_connection", { request: input });
}

/** 更新连接(整体替换编辑字段)。 */
export function updateConnection(
  id: string,
  input: ConnectionInput,
): Promise<ConnectionDto> {
  return invokeUnwrapped("update_connection", { request: { id, input } });
}

/** 删除连接(调用方负责二次确认)。 */
export function deleteConnection(id: string): Promise<void> {
  return invokeUnwrapped("delete_connection", { request: { id } });
}

/** 克隆连接,名称追加"(副本)"。 */
export function duplicateConnection(id: string): Promise<ConnectionDto> {
  return invokeUnwrapped("duplicate_connection", { request: { id } });
}

/** 移动连接到目标分组;targetGroupId 传 null 移到根目录,position 缺省追加末尾。 */
export function moveConnection(
  id: string,
  targetGroupId: string | null,
  position?: number,
): Promise<void> {
  return invokeUnwrapped("move_connection", {
    request: { id, targetGroupId, position: position ?? null },
  });
}

/** 新建分组;parentId 传 null 建在根级。 */
export function createGroup(name: string, parentId?: string | null): Promise<GroupDto> {
  return invokeUnwrapped("create_group", {
    request: { name, parentId: parentId ?? null },
  });
}

/** 重命名分组。 */
export function renameGroup(id: string, name: string): Promise<void> {
  return invokeUnwrapped("rename_group", { request: { id, name } });
}

/** 删除分组;策略决定非空时拒绝或子项提升到父级。 */
export function deleteGroup(id: string, mode: DeleteGroupMode): Promise<void> {
  return invokeUnwrapped("delete_group", { request: { id, mode } });
}

/** 移动分组;目标父级必须是根级。 */
export function moveGroup(
  id: string,
  targetParentId: string | null,
  position?: number,
): Promise<void> {
  return invokeUnwrapped("move_group", {
    request: { id, targetParentId, position: position ?? null },
  });
}

File: src-tauri/src/commands/connections.rs
//! 连接 command:边界校验 → 状态取服务 → 用例 → DTO 映射(AGENTS.md §4.2)。

use tauri::State;

use crate::application::connections::ConnectionError;
use crate::dto::common::IpcResult;
use crate::dto::connection::{
    parse_conn_id, parse_group_id, ConnectionDto, ConnectionInputDto, ConnectionNodeDto,
    IdRequestDto, MoveConnectionRequestDto, UpdateConnectionRequestDto,
};
use crate::shared::error::{IpcError, IpcErrorCode};
use crate::state::AppState;

/// 连接管理应用错误 → IPC 错误码。
pub(super) fn to_ipc_error(err: ConnectionError) -> IpcError {
    match err {
        ConnectionError::Domain(_)
        | ConnectionError::GroupNotEmpty
        | ConnectionError::NestingTooDeep => {
            IpcError::new(IpcErrorCode::InvalidArgument, err.to_string())
        }
        ConnectionError::NotFound(_) => IpcError::new(IpcErrorCode::NotFound, err.to_string()),
        ConnectionError::Storage(e) => IpcError::new(IpcErrorCode::Internal, e.to_string()),
        ConnectionError::Secret(e) => {
            IpcError::new(IpcErrorCode::KeyringUnavailable, e.to_string())
        }
    }
}

/// 列出连接树(分组在前、连接在后,同级按 position/name 排序)。
#[tauri::command]
pub fn list_connections(state: State<AppState>) -> IpcResult<Vec<ConnectionNodeDto>> {
    match state.connections.list_tree() {
        Ok(tree) => IpcResult::ok(tree.into_iter().map(Into::into).collect()),
        Err(err) => IpcResult::err(to_ipc_error(err)),
    }
}

/// 新建连接。
#[tauri::command]
pub fn create_connection(
    state: State<AppState>,
    request: ConnectionInputDto,
) -> IpcResult<ConnectionDto> {
    let draft = match request.to_draft() {
        Ok(draft) => draft,
        Err(err) => return IpcResult::err(to_ipc_error(err)),
    };
    match state.connections.create_connection(draft) {
        Ok(config) => IpcResult::ok(config.into()),
        Err(err) => IpcResult::err(to_ipc_error(err)),
    }
}

/// 更新连接(整体替换编辑字段)。
#[tauri::command]
pub fn update_connection(
    state: State<AppState>,
    request: UpdateConnectionRequestDto,
) -> IpcResult<ConnectionDto> {
    let result = parse_conn_id(request.id)
        .and_then(|id| Ok((id, request.input.to_draft()?)))
        .and_then(|(id, draft)| state.connections.update_connection(&id, draft));
    match result {
        Ok(config) => IpcResult::ok(config.into()),
        Err(err) => IpcResult::err(to_ipc_error(err)),
    }
}

/// 删除连接(前端负责二次确认)。
#[tauri::command]
pub fn delete_connection(state: State<AppState>, request: IdRequestDto) -> IpcResult<()> {
    match parse_conn_id(request.id).and_then(|id| state.connections.delete_connection(&id)) {
        Ok(()) => IpcResult::ok(()),
        Err(err) => IpcResult::err(to_ipc_error(err)),
    }
}

/// 克隆连接,名称追加"(副本)"。
#[tauri::command]
pub fn duplicate_connection(
    state: State<AppState>,
    request: IdRequestDto,
) -> IpcResult<ConnectionDto> {
    match parse_conn_id(request.id).and_then(|id| state.connections.duplicate_connection(&id)) {
        Ok(config) => IpcResult::ok(config.into()),
        Err(err) => IpcResult::err(to_ipc_error(err)),
    }
}

/// 移动连接到目标分组与位置。
#[tauri::command]
pub fn move_connection(state: State<AppState>, request: MoveConnectionRequestDto) -> IpcResult<()> {
    let result = parse_conn_id(request.id).and_then(|id| {
        let target = request.target_group_id.map(parse_group_id).transpose()?;
        state
            .connections
            .move_connection(&id, target, request.position)
    });
    match result {
        Ok(()) => IpcResult::ok(()),
        Err(err) => IpcResult::err(to_ipc_error(err)),
    }
}
